### ISSUE TITLE
Migration failed: Maximum call stack size exceeded on `@angular-eslint/schematics` 2.1.1 to 3.0.0

### ISSUE THREAD

**Issue by KeithGillette** [angular-eslint/angular-eslint#416]
Running ` npx ng update @angular-eslint/schematics` on existing `@angular-eslint/*@2.1.1` project produces:
```
Using package manager: 'npm'
Collecting installed dependencies...
Found 112 dependencies.
Fetching dependency metadata from registry...
    Updating package.json with dependency @angular-eslint/schematics @ "3.0.0" (was "2.1.1")...
  UPDATE package.json (12795 bytes)
✔ Packages installed successfully.
** Executing migrations of package '@angular-eslint/schematics' **

❯ Updates @angular-eslint to v3.
✖ Migration failed: Maximum call stack size exceeded
  See "/private/var/folders/t5/nv32j92n7rngpy281m6gj8n00000gn/T/ng-pRL8KS/angular-errors.log" for further details.

```
**angular-errors.log:**
```
[error] RangeError: Maximum call stack size exceeded
    at /Path/to/node_modules/@angular-eslint/schematics/dist/utils.js:116:13
    at Array.forEach (<anonymous>)
    at allFilesInDirInHost (/Path/to/node_modules/@angular-eslint/schematics/dist/utils.js:115:17)
    at /Path/to/node_modules/@angular-eslint/schematics/dist/utils.js:116:21
    at Array.forEach (<anonymous>)
    at allFilesInDirInHost (/Path/to/node_modules/@angular-eslint/schematics/dist/utils.js:115:17)
    at Object.getAllSourceFilesForProject (/Path/to/node_modules/@angular-eslint/schematics/dist/utils.js:131:12)
    at applyRecommendedExtraExtends (/Path/to/node_modules/@angular-eslint/schematics/dist/migrations/update-3-0-0/update-3-0-0.js:58:50)
    at MergeMapSubscriber.project (/Path/to/node_modules/@angular-devkit/schematics/src/rules/call.js:75:24)
    at MergeMapSubscriber._tryNext (/Path/to/node_modules/@angular-devkit/schematics/node_modules/rxjs/internal/operators/mergeMap.js:67:27)
    at MergeMapSubscriber._next (/Path/to/node_modules/@angular-devkit/schematics/node_modules/rxjs/internal/operators/mergeMap.js:57:18)
    at MergeMapSubscriber.Subscriber.next (/Path/to/node_modules/@angular-devkit/schematics/node_modules/rxjs/internal/Subscriber.js:66:18)
    at MergeMapSubscriber.notifyNext (/Path/to/node_modules/@angular-devkit/schematics/node_modules/rxjs/internal/operators/mergeMap.js:93:26)
    at SimpleInnerSubscriber._next (/Path/to/node_modules/@angular-devkit/schematics/node_modules/rxjs/internal/innerSubscribe.js:27:21)
    at SimpleInnerSubscriber.Subscriber.next (/Path/to/node_modules/@angular-devkit/schematics/node_modules/rxjs/internal/Subscriber.js:66:18)
    at Observable._subscribe (/Path/to/node_modules/@angular-devkit/schematics/node_modules/rxjs/internal/util/subscribeToArray.js:5:20)
```

**eslint.json:**
```
{
  "root": true,
  "ignorePatterns": [
    "projects/**/*"
  ],
  "overrides": [
    {
      "files": [
        "*.ts"
      ],
      "parserOptions": {
        "project": [
          "tsconfig.json"
        ]
      },
      "extends": [
        "plugin:@angular-eslint/ng-cli-compat",
        "plugin:@angular-eslint/ng-cli-compat--formatting-add-on",
        "plugin:@angular-eslint/template/process-inline-templates"
      ],
      "plugins": [
        "eslint-plugin-unicorn",
        "eslint-plugin-rxjs"
      ],
      "rules": {
        "@angular-eslint/component-max-inline-declarations": "error",
        "@angular-eslint/component-selector": [
          "error",
          {
            "type": "element",
            "prefix": "tt",
            "style": "kebab-case"
          }
        ],
        "@angular-eslint/directive-selector": [
          "error",
          {
            "type": "attribute",
            "prefix": "tt",
            "style": "camelCase"
          }
        ],
        "@angular-eslint/no-attribute-decorator": "error",
        "@angular-eslint/no-forward-ref": "error",
        "@angular-eslint/no-input-rename": "error",
        "@angular-eslint/no-lifecycle-call": "error",
        "@angular-eslint/no-pipe-impure": "error",
        "@angular-eslint/pipe-prefix": [
          "error",
          {
            "prefixes": [
              "tt"
            ]
          }
        ],
        "@angular-eslint/prefer-output-readonly": "error",
        "@angular-eslint/use-component-selector": "error",
        "@angular-eslint/use-injectable-provided-in": "error",
        "@typescript-eslint/array-type": [
          "error",
          {
            "default": "array"
          }
        ],
        "@typescript-eslint/consistent-type-definitions": "error",
        "@typescript-eslint/dot-notation": "off",
        "@typescript-eslint/explicit-member-accessibility": [
          "error",
          {
            "accessibility": "explicit"
          }
        ],
        // @ToDo: Update defaults to match project usage & clean up project, or turn off
        "@typescript-eslint/member-ordering": "warn",
        "@typescript-eslint/naming-convention": [
          "error",
          {
            "selector": "parameter",
            "format": [
              "camelCase",
              "PascalCase"
            ],
            "leadingUnderscore": "allow",
            "trailingUnderscore": "allow"
          },
          {
            "selector": "enumMember",
            "format": [
              "camelCase",
              "PascalCase"
            ]
          }
        ],
        // DTO interfaces use type transformations on classes without adding properties
        "@typescript-eslint/no-empty-interface": "off",
        // Angular Modules are often decorated empty classes
        "@typescript-eslint/no-extraneous-class": "off",
        "@typescript-eslint/no-for-in-array": "error",
        "@typescript-eslint/no-inferrable-types": "off",
        "@typescript-eslint/no-shadow": [
          "error",
          {
            "hoist": "never"
          }
        ],
        "@typescript-eslint/no-this-alias": "error",
        "@typescript-eslint/no-unused-expressions": [
          "error",
          {
            "allowShortCircuit": true,
            "allowTernary": true
          }
        ],
        "@typescript-eslint/quotes": [
          "error",
          "single",
          {
            "avoidEscape": true
          }
        ],
        "@typescript-eslint/require-await": "error",
        "@typescript-eslint/unbound-method": [
          "error",
          {
            "ignoreStatic": true
          }
        ],
        // Just a style preference to override [default](https://eslint.org/docs/rules/arrow-body-style),
        "arrow-body-style": "off",
        "arrow-parens": [
          "error",
          "always"
        ],
        "brace-style": [
          "error",
          "1tbs"
        ],
        "id-blacklist": "off",
        "id-match": "off",
        "import/no-default-export": "error",
        "import/no-deprecated": "error",
        "import/no-extraneous-dependencies": "off",
        "max-len": [
          "error",
          {
            "code": 188,
            "ignoreComments": true,
            "ignoreUrls": true,
            "ignoreStrings": true,
            "ignoreTemplateLiterals": true,
            "ignoreRegExpLiterals": true
          }
        ],
        "no-duplicate-case": "error",
        "no-duplicate-imports": "error",
        "no-extra-bind": "error",
        "no-irregular-whitespace": "error",
        "no-new-func": "error",
        "no-redeclare": "error",
        "no-restricted-syntax": [
          "error",
          "ForInStatement"
        ],
        "no-return-await": "error",
        "no-sequences": "error",
        "no-template-curly-in-string": "error",
        "no-underscore-dangle": "off",
        // Dependency Injection requires constructor
        "no-useless-constructor": "off",
        // https://eslint.org/docs/rules/object-shorthand#require-object-literal-shorthand-syntax-object-shorthand
        "object-shorthand": [
          "error",
          "consistent"
        ],
        "prefer-template": "error",
        "rxjs/no-async-subscribe": "error",
        "rxjs/no-create": "error",
        "rxjs/no-ignored-replay-buffer": "error",
        "rxjs/no-index": "error",
        "rxjs/no-internal": "error",
        "rxjs/no-nested-subscribe": "error",
        "rxjs/no-sharereplay": [
          "error",
          {
            "allowConfig": true
          }
        ],
        "rxjs/no-subject-unsubscribe": "error",
        "rxjs/no-unbound-methods": "error",
        "rxjs/no-unsafe-subject-next": "error",
        "rxjs/no-unsafe-takeuntil": "error",
        "space-in-parens": [
          "off",
          "never"
        ],
        "unicorn/filename-case": "error"
      }
    },
    {
      "files": [
        "*.html"
      ],
      "extends": [
        "plugin:@angular-eslint/template/recommended"
      ],
      "rules": {
        "@angular-eslint/template/conditional-complexity": "error",
        "@angular-eslint/template/cyclomatic-complexity": [
          "error",
          {
            "maxComplexity": 5
          }
        ]
      }
    }
  ]
}
```

Troubleshooting recommendations?

**Comment by JamesHenry** [comment]
Hmm interesting @KeithGillette...

`allFilesInDirInHost` in the utils.js file is clearly the issue, it is a recursive function and something about your workspace directory structure is causing it to hit a Maximum call stack.

I presume it's not possible to give me temporary access to the repo in question?

If that is the case, the only thing I can recommend is trying to debug/add logging to that function as it runs and figure out where it is getting stuck.

Many thanks!

**Comment by JamesHenry** [comment]
Wait sorry I see you opened effectively this same issue before: https://github.com/angular-eslint/angular-eslint/issues/295

AFAIK you are the only person who has ever run into this, so would be very curious to know about your workspace

**Comment by KeithGillette** [comment]
Thanks for the reply, @JamesHenry. Yes, I got this previous error on an initial attempt to migrate this same monorepository project from `tslint` to `eslint`, but the issue resolved with a later release of the schematic, as I recall. And yes, our project is private, so I can't really give you access, but I'll try to describe its idiosyncratic legacy layout in hopes it may provide some clues. Here's the essential file structure, with ellipses representing omissions (like source code) and items in square brackets providing explanatory notes:
```
- clients/
  - browser/
    - app/
      - …
      - shared~server/ [symlink to ../../shared]
    - .eslintrc.json
    - tsconfig.app.json
- server/
  - server/
    -  …
    - shared~client/ [symlink to ../../shared]
  - .eslintrc.json
  - tsconfig.json
  - package.json
  - node_modules/…
- shared/…
- zapier/
  - src/…
  - .eslintrc.json
  - package.json
  - tsconfig.json
  - node_modules/
- angular.json
- .eslintrc.json
- package.json
- tsconfig.json
- node_modules/…
- …
```

Each of the 3 nested `.eslintrc.json` extends from the top-level `.eslintrc.json` provided previously, providing overrides for for each of the 3 projects defined in `angular.json`:
```
projects: {
  browser: {
    root: "clients/browser",
    sourceRoot: "clients/browser",
  },
  server: {
    root: "server",
    sourceRoot: "server/server",
  },
  zapier: {
    root: "zapier",
    sourceRoot: "zapier/src",  
  },
}
```

How do you recommend running the migration schematic with logging to debug? When I execute `ng g @angular-eslint/schematics/dist/migrations:update-3-0-0`, the doesn't stop with the same `Maximum call stack size exceeded` error, but seems to get stuck endlessly looping through a subset of temporary build output dependency directories in `/server/.meteor/local/.build-garbage-ngtnd3.a26m8/programs/server/npm/node_modules/`.

**Comment by KeithGillette** [comment]
Tried with v4.0.0 release, but get the exact same error:
```
✔ Packages installed successfully.
** Executing migrations of package '@angular-eslint/schematics' **

❯ Updates @angular-eslint to v3.
✖ Migration failed: Maximum call stack size exceeded
```

**Comment by KeithGillette** [comment]
Given the hint I had from the infinite loop I observed when running `ng g @angular-eslint/schematics/dist/migrations:update-3-0-0`, I decided to run the migration again after deleting all repository build caches, and it worked properly, so I'm closing this issue. However, I don't understand why the schematic would need to descend the build directories in the first place.

**Comment by JamesHenry** [comment]
@KeithGillette In #425 I have refactored things within schematics and taken the opportunity to add some logic to take `.gitignore` data into account when figuring out what files to iterate through as part of the conversion process.

I believe this should fix the issue you encountered, it will be released in `4.1.0`

**Comment by KeithGillette** [comment]
Thanks, @JamesHenry. As I said, I was able to get the previous schematics migration to run after deleting the build caches. I ran `npx ng update @angular-eslint/schematics` without deleting build caches in our project this time and it worked properly, but I don't know if that's a fair test since there are probably fewer caches than previously and the migration from `2.1.1` to `3.0.0` had already occurred. Follow-up greatly appreciated regardless.
